tests/formats/smi.cpp

```diff
@@ -255,27 +251,28 @@ TEST_CASE("Check parsing results") {
     }
 }
 
-// To use in loops in order to iterate over files in a specific directory.
-struct directory_files_iterator {
-    typedef fs::recursive_directory_iterator iterator;
-    directory_files_iterator(fs::path p) : p_(p) {}
+TEST_CASE("Errors in SMI format") {
+    auto file = Trajectory("data/smi/bad/bad_element.smi");
+    CHECK_THROWS_WITH(file.read(), "SMI Reader: bare non-organic atom: 'W'");
 
-    iterator begin() { return fs::recursive_directory_iterator(p_); }
-    iterator end() { return fs::recursive_directory_iterator(); }
+    file = Trajectory("data/smi/bad/bad_paren.smi");
+    CHECK_THROWS_WITH(file.read(), "SMI Reader: 1 unclosed '('(s)");
 
-    fs::path p_;
-};
+    file = Trajectory("data/smi/bad/bad_percentage_sign.smi");
+    CHECK_THROWS_WITH(file.read(), "SMI Reader: rings defined with '%' must be double digits");
 
-TEST_CASE("Errors in SMI format") {
-    for (auto entry : directory_files_iterator("data/smi/bad/")) {
-        auto test = [=](){
-            // We can throw either when creating the trajectory, or when reading
-            // the frame, depending on the type of error
-            auto file = Trajectory(entry.path().string());
-            file.read();
-        };
-        CHECK_THROWS(test());
-    }
+    file = Trajectory("data/smi/bad/bad_ring.smi");
+    CHECK_THROWS_WITH(file.read(), "SMI Reader: unclosed ringid '4'");
+
+    file = Trajectory("data/smi/bad/bad_symbol.smi");
+    CHECK_THROWS_WITH(file.read(), "SMI Reader: unknown symbol: '`'");
+
+    file = Trajectory("data/smi/bad/misplaced_property.smi");
+    CHECK_THROWS_WITH(file.read(), "SMI Reader: symbol not allowed outside of property: '@'");
+
+    auto bad = std::string("C)");
+    file = Trajectory::memory_reader(bad.c_str(), bad.size(), "SMI");
+    CHECK_THROWS_WITH(file.read(), "SMI Reader: unmatched ')'");
```

Comment:
Off-topic, but would an additional overload of `Trajectory::memory_reader` with `std::string` be useful?

Reply:
That, or with a `span<const char>`. Even better if we can get automatic conversions from both `std::string` and `std::vector<char>` =)